internal/controller/alert_rule.go

```diff
@@ -91,61 +88,85 @@ func parseRuleGroups(promRuleName, tier, service string, in []monitoringv1.RuleG
+// a slice of []monitoringv1.Rule is returned as the result would be multiple
+// absent metric alert rules (one for each time series).
+func ParseAlertRule(tier, service string, in monitoringv1.Rule) ([]monitoringv1.Rule, error) {
+	exprStr := in.Expr.String()
+	mex := &metricNameExtractor{expr: exprStr, found: map[string]struct{}{}}
+	exprNode, err := parser.ParseExpr(exprStr)
+	if err == nil {
+		err = parser.Walk(mex, exprNode, nil)
+	}
+	if err != nil {
+		// The returned error has the expression in last because
+		// it could contain newline chracters.
+		return nil, fmt.Errorf("could not parse rule expression: %s: %s", err.Error(), exprStr)
+	}
+	if len(mex.found) == 0 {
+		return nil, nil
+	}
+
 	// Carry over labels from the original alert
-	if originalLabels != nil {
-		if v, ok := originalLabels["tier"]; ok {
-			if !strings.Contains(v, "$labels") {
-				tier = v
-			}
+	if origLab := in.Labels; origLab != nil {
+		if v, ok := origLab["tier"]; ok && !strings.Contains(v, "$labels") {
+			tier = v
 		}
-		if v, ok := originalLabels["service"]; ok {
-			if !strings.Contains(v, "$labels") {
-				service = v
-			}
+		if v, ok := origLab["service"]; ok && !strings.Contains(v, "$labels") {
+			service = v
 		}
 	}
-	labels := map[string]string{
+	lab := map[string]string{
 		"tier":     tier,
 		"service":  service,
 		"severity": "info",
 	}
 
-	// Generate an alert name from metric name:
-	//   network:tis_a_metric:rate5m -> Absent("tier")("service")NetworkTisAMetricRate5m
-	words := []string{"absent", tier, service}
-	sL1 := strings.Split(metric, "_")
-	for _, v := range sL1 {
-		sL2 := strings.Split(v, ":")
-		words = append(words, sL2...)
+	// Sort metric names alphabetically for consistent test results.
+	metrics := make([]string, 0, len(mex.found))
+	for k := range mex.found {
+		metrics = append(metrics, k)
 	}
-	// Avoid name stuttering
-	var alertName string
-	var prev string
-	for _, v := range words {
-		l := strings.ToLower(v)
-		if l != prev {
-			prev = l
-			alertName += strings.Title(l)
+	sort.Strings(metrics)
+
+	out := make([]monitoringv1.Rule, 0, len(metrics))
+	for _, m := range metrics {
+		// Generate an alert name from metric name:
+		//   network:tis_a_metric:rate5m -> AbsentTierServiceNetworkTisAMetricRate5m
+		words := []string{"absent", tier, service}
+		sL1 := strings.Split(m, "_")
+		for _, v := range sL1 {
+			sL2 := strings.Split(v, ":")
+			words = append(words, sL2...)
+		}
+		// Avoid name stuttering
+		var alertName string
+		var prev string
+		for _, v := range words {
+			l := strings.ToLower(v)
+			if prev != l {
+				prev = l
+				alertName += strings.Title(l)
+			}
 		}
-	}
 
-	annotations := map[string]string{
-		"summary":     fmt.Sprintf("missing %s", metric),
-		"description": fmt.Sprintf("The metric '%s' is missing", metric),
-	}
+		ann := map[string]string{
+			"summary":     fmt.Sprintf("missing %s", m),
+			"description": fmt.Sprintf("The metric '%s' is missing", m),
```

Comment:
The description should also hint at why this is a problem.
```suggestion
			"description": fmt.Sprintf("The metric '%s' is missing. Alerts using it may not fire as intended.", m),
```

Reply:
Done.